repose-aggregator/components/filters/merge-header/src/main/scala/org/openrepose/filters/mergeheader/MergeHeaderFilter.scala

```diff
@@ -1,39 +1,100 @@
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
 import org.openrepose.core.filter.FilterConfigHelper
-import org.openrepose.core.filter.logic.impl.FilterLogicHandlerDelegate
 import org.openrepose.core.services.config.ConfigurationService
 
+import scala.collection.JavaConversions._
+
 @Named
-class MergeHeaderFilter @Inject()(configurationService: ConfigurationService) extends Filter with LazyLogging {
+class MergeHeaderFilter @Inject()(configurationService: ConfigurationService)
+  extends Filter with UpdateListener[MergeHeaderConfig] with LazyLogging {
+
   private final val DEFAULT_CONFIG = "merge-header.cfg.xml"
 
-  private var config: String = _
-  private var handlerFactory: MergeHeaderHandlerFactory = _
+  private var initialized = false
+  private var configFilename: String = _
+  private var filterConfig: MergeHeaderConfig = _
 
   override def init(filterConfig: FilterConfig): Unit = {
-    config = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
-    logger.info(s"Initializing MergeHeaderFilter using config $config")
-    handlerFactory = new MergeHeaderHandlerFactory()
+    configFilename = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing MergeHeaderFilter using config $configFilename")
     val xsdURL = getClass.getResource("/META-INF/config/schema/merge-header.xsd")
-    // TODO: Clean up the asInstanceOf below, if possible?
-    configurationService.subscribeTo(filterConfig.getFilterName,
-      config,
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      configFilename,
       xsdURL,
-      handlerFactory.asInstanceOf[UpdateListener[MergeHeaderConfig]],
-      classOf[MergeHeaderConfig])
+      this,
+      classOf[MergeHeaderConfig]
+    )
   }
 
   override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
-    new FilterLogicHandlerDelegate(servletRequest, servletResponse, filterChain).doFilter(handlerFactory.newHandler)
+    if (!isInitialized) {
+      logger.error("Filter has not yet initialized...")
+      servletResponse.asInstanceOf[HttpServletResponse].sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
+    } else {
+      val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+      val httpServletResponse = servletResponse.asInstanceOf[HttpServletResponse]
+
+      Option(filterConfig.getRequest) foreach { requestConfig =>
+        requestConfig.getHeader foreach { name =>
+          val value = wrappedRequest.getHeaders(name).mkString(", ")
+          if (value.nonEmpty) {
+            logger.debug(s"REQUEST: merging header $name to $value")
+            wrappedRequest.replaceHeader(name, value)
+          }
+        }
+      }
+
+      filterChain.doFilter(wrappedRequest, servletResponse)
+
+      Option(filterConfig.getResponse).foreach { responseConfig =>
```

Comment:
Only because I already left another comment...   that period before foreach makes me a sad panda since the previous foreach doesn't have one.